dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/QueryItemHelper.java

```diff
@@ -117,4 +132,98 @@ else if ( item.hasOptionSet() && (option = item.getOptionSet().getOptionByCode(
             return value + itemValue;
         }
     }
+
+    /**
+     * Returns a map of metadata item options and {@link Option}.
+     *
+     * @param grid the Grid instance.
+     * @param params the EventQueryParams.
+     * @return a list of options.
+     */
+    public static List<Option> getItemOptions( final Grid grid, final EventQueryParams params )
+    {
+        final List<Option> options = new ArrayList<>();
+
+        for ( int i = 0; i < grid.getHeaders().size(); ++i )
+        {
+            final GridHeader gridHeader = grid.getHeaders().get( i );
+
+            if ( gridHeader.hasOptionSet() && isNotEmpty( grid.getRows() ) )
+            {
+                options.addAll( getItemOptionsThatMatchesRows( grid, i, gridHeader ) );
+            }
+            else if ( gridHeader.hasOptionSet() && isEmpty( grid.getRows() ) )
+            {
+                options.addAll( getItemOptionsWhenRowsIsEmpty( params ) );
+            }
+        }
+
+        return options.stream().distinct().collect( toList() );
+    }
+
+    private static List<Option> getItemOptionsWhenRowsIsEmpty( final EventQueryParams params )
```

Comment:
Small typo here: AreEmpty, not IsEmpty. Maybe we can find a more succinct name for the method too, like getItemOptionsNoRows.